jetty-hazelcast/src/main/java/org/eclipse/jetty/hazelcast/session/HazelcastSessionDataStore.java

```diff
@@ -0,0 +1,211 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2017 Mort Bay Consulting Pty. Ltd.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.hazelcast.session;
+
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import org.eclipse.jetty.server.session.AbstractSessionDataStore;
+import org.eclipse.jetty.server.session.SessionContext;
+import org.eclipse.jetty.server.session.SessionData;
+import org.eclipse.jetty.server.session.SessionDataMap;
+import org.eclipse.jetty.server.session.SessionDataStore;
+import org.eclipse.jetty.util.annotation.ManagedObject;
+import org.eclipse.jetty.util.component.AbstractLifeCycle;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Session data stored in Hazelcast
+ */
+@ManagedObject
+public class HazelcastSessionDataStore
+    extends AbstractSessionDataStore
+    implements SessionDataStore
+{
+
+    private  final static Logger LOG = Log.getLogger( "org.eclipse.jetty.server.session");
+
+    private HazelcastInstance hazelcastInstance;
+
+    private String jettySessionMapName;
+
+    private IMap<String, SessionData> sessionDataMap;
+
+    public HazelcastSessionDataStore()
+    {
+        // no op
+    }
+
+    public HazelcastSessionDataStore( HazelcastInstance hazelcastInstance, String jettySessionMapName )
+    {
+        this.hazelcastInstance = hazelcastInstance;
+        this.jettySessionMapName = jettySessionMapName;
+        this.sessionDataMap = hazelcastInstance.getMap( getJettySessionMapName() );
+    }
+
+    public HazelcastInstance getHazelcastInstance()
+    {
+        return hazelcastInstance;
+    }
+
+    public void setHazelcastInstance( HazelcastInstance hazelcastInstance )
+    {
+        this.hazelcastInstance = hazelcastInstance;
+    }
+
+    public String getJettySessionMapName()
+    {
+        return jettySessionMapName;
+    }
+
+    public void setJettySessionMapName( String jettySessionMapName )
+    {
+        this.jettySessionMapName = jettySessionMapName;
+    }
+
+
+    @Override
+    public SessionData load( String id )
+        throws Exception
+    {
```

Comment:
Take a look at the load method InfinispanSessionDataStore - we ensure we run it in the scope of a the webapp's classloader with _context.run(Runnable), and we also use atomics to make sure that 2 or more simultaneous requests only result in a single load.